src/NuGet.Core/NuGet.PackageManagement/Resolution/PrunePackageTree.cs

```diff
@@ -63,30 +63,33 @@ private static void CollectAllowedFromDependenciesOfPackagesToInstall(IEnumerabl
         {
             var prereleasePackageToInstall = new HashSet<PackageIdentity>(packagesToInstall.Where(p => p.HasVersion && p.Version.IsPrerelease), PackageIdentityComparer.Default);
 
-            var maxDepth = packages.Select(p => p.Id).Distinct(StringComparer.OrdinalIgnoreCase).Count();
+            var visitedNodes = new HashSet<SourcePackageDependencyInfo>();
+            var packagesLookup = packages.ToLookup(p => p.Id);
 
             foreach (var packageToInstall in packages.Where(p => prereleasePackageToInstall.Contains(p)))
             {
-                int depth = 0;
-                WalkDependencies(packages, packageToInstall, allowed, depth, maxDepth);
+                WalkDependencies(packagesLookup, packageToInstall, allowed, visitedNodes);
             }
         }
 
-        private static void WalkDependencies(IEnumerable<SourcePackageDependencyInfo> packages, SourcePackageDependencyInfo packageToInstall, HashSet<string> allowed, int depth, int maxDepth)
+        private static void WalkDependencies(ILookup<string, SourcePackageDependencyInfo> packagesLookup, SourcePackageDependencyInfo packageToInstall, HashSet<string> allowed, HashSet<SourcePackageDependencyInfo> visitedNodes)
         {
-            if (depth > maxDepth)
-            {
-                return;
-            }
+            var queue = new Queue<SourcePackageDependencyInfo>();
+            queue.Enqueue(packageToInstall);
 
-            if (packageToInstall.Dependencies != null)
+            while (queue.Count > 0)
             {
-                foreach (var dependency in packageToInstall.Dependencies)
+                var package = queue.Dequeue();
+                foreach (var dependency in package.Dependencies)
                 {
                     allowed.Add(dependency.Id);
-                    foreach (SourcePackageDependencyInfo dependentPackage in packages.Where(p => p.Id.Equals(dependency.Id, StringComparison.OrdinalIgnoreCase)))
+                    var packages = packagesLookup[dependency.Id];
```

Comment:
This can probably hit a missing dependency, this should be a dictionary and then use TryGet